bpo-29738: Fix memory leak in _get_crl_dp

olivielpeau opened the pull request:
* Remove conditional on free of `dps`, since `dps` is now allocated for
all versions of OpenSSL
* Use `sk_DIST_POINT_pop_free` instead of `sk_DIST_POINT_free` since
the latter doesn't free the individual elements of the stack
* Remove call to `x509_check_ca` since it was only used to cache
the `crldp` field of the certificate

olivielpeau committed:
bpo-29738: Fix memory leak in _get_crl_dp

* Remove conditional on free of `dps`, since `dps` is now allocated for
all versions of OpenSSL
* Use `sk_DIST_POINT_pop_free` instead of `sk_DIST_POINT_free` since
the latter doesn't free the individual elements of the stack
* Remove call to `x509_check_ca` since it was only used to cache
the `crldp` field of the certificate

the-knights-who-say-ni commented:
Hello, and thanks for your contribution!

I'm a bot set up to make sure that the project can legally accept your contribution by verifying you have signed the [PSF contributor agreement](https://www.python.org/psf/contrib/contrib-form/) (CLA).

Unfortunately our records indicate you have not signed the CLA. For legal reasons we need you to sign this before we can look at your contribution. Please follow these steps to rectify the issue:

1. Sign the [PSF contributor agreement](https://www.python.org/psf/contrib/contrib-form/). The "bugs.python.org username" requested by the form is the "Login name" field in ["Your Details" at b.p.o](https://cloud.githubusercontent.com/assets/2680980/23276970/d14a380c-f9d1-11e6-883d-e13b6b211239.png)
2. **Wait at least one US business day** and then check the "Contributor form received entry under "Your Details" on [bugs.python.org](https://bugs.python.org/) to see if your account has been marked as having signed the CLA (the delay is due to a person having to manually check your signed CLA)
3. Reply here saying you have completed the above steps


Thanks again to your contribution and we look forward to looking at it!

olivielpeau commented:
@the-knights-who-say-ni : CLA signed

alex reviewed (approved):
This looks correct to me.

I can't remember the rules for merging and backports, so I'm hoping another core dev will actually land :-)

tiran commented on Modules/_ssl.c:
```diff
@@ -1257,9 +1253,7 @@ _get_crl_dp(X509 *certificate) {
-#if OPENSSL_VERSION_NUMBER < 0x10001000L
-    sk_DIST_POINT_free(dps);
-#endif
+    sk_DIST_POINT_pop_free(dps, DIST_POINT_free);
```
```
CRL_DIST_POINTS_free(dps);
```

tiran reviewed (changes requested):
CRL_DIST_POINTS_free is available in all supported versions of OpenSSL (recent 0.9.8+) and LibreSSL.

olivielpeau committed:
bpo-29738: Use CRL_DIST_POINTS_free

Addresses PR review.

CRL_DIST_POINTS_free is available in all supported versions of OpenSSL
(recent 0.9.8+) and LibreSSL.

olivielpeau commented:
@tiran Thanks for the review! I've addressed your comment, let me know if this looks good now

Mariatta commented:
Thanks @olivielpeau and congrats for your first contribution to CPython :tada:
